File: api-comparison/python/common/repository.py
"""Repositório com TODA a lógica de negócio (CRUD + consultas do enunciado).

Cada servidor (SOAP/REST/GraphQL/gRPC) é apenas um adaptador fino sobre estas
funções. Nenhuma regra de negócio vive na camada de protocolo.

Convenção: as funções retornam dicionários simples (DTOs) para que cada
protocolo serialize do seu jeito sem acoplar ao ORM.
"""
from typing import Optional
from sqlalchemy import select, delete, func
from sqlalchemy.orm import Session
from .db import SessionLocal, User, Music, Playlist, PlaylistMusic


# ---------- serialização para DTO ----------
def _user_dto(u: User) -> dict:
    return {"id": u.id, "name": u.name, "email": u.email}


def _music_dto(m: Music) -> dict:
    return {
        "id": m.id, "title": m.title, "artist": m.artist,
        "album": m.album, "duration_seconds": m.duration_seconds,
    }


def _playlist_dto(p: Playlist) -> dict:
    return {"id": p.id, "name": p.name, "user_id": p.user_id}


class NotFound(Exception):
    pass


# =================== USERS ===================
def create_user(name: str, email: str) -> dict:
    with SessionLocal() as s:
        u = User(name=name, email=email)
        s.add(u); s.commit(); s.refresh(u)
        return _user_dto(u)


def get_user(user_id: int) -> dict:
    with SessionLocal() as s:
        u = s.get(User, user_id)
        if not u:
            raise NotFound(f"user {user_id}")
        return _user_dto(u)


def list_users(limit: int = 100, offset: int = 0) -> list[dict]:
    with SessionLocal() as s:
        rows = s.scalars(
            select(User).order_by(User.id).limit(limit).offset(offset)
        ).all()
        return [_user_dto(u) for u in rows]


def update_user(user_id: int, name: Optional[str], email: Optional[str]) -> dict:
    with SessionLocal() as s:
        u = s.get(User, user_id)
        if not u:
            raise NotFound(f"user {user_id}")
        if name is not None:
            u.name = name
        if email is not None:
            u.email = email
        s.commit(); s.refresh(u)
        return _user_dto(u)


def delete_user(user_id: int) -> bool:
    with SessionLocal() as s:
        u = s.get(User, user_id)
        if not u:
            return False
        s.delete(u); s.commit()
        return True


# =================== MUSICS ===================
def create_music(title: str, artist: str, album: Optional[str],
                 duration_seconds: int) -> dict:
    with SessionLocal() as s:
        m = Music(title=title, artist=artist, album=album,
                  duration_seconds=duration_seconds)
        s.add(m); s.commit(); s.refresh(m)
        return _music_dto(m)


def get_music(music_id: int) -> dict:
    with SessionLocal() as s:
        m = s.get(Music, music_id)
        if not m:
            raise NotFound(f"music {music_id}")
        return _music_dto(m)


def list_musics(limit: int = 100, offset: int = 0) -> list[dict]:
    with SessionLocal() as s:
        rows = s.scalars(
            select(Music).order_by(Music.id).limit(limit).offset(offset)
        ).all()
        return [_music_dto(m) for m in rows]


def update_music(music_id: int, title: Optional[str], artist: Optional[str],
                 album: Optional[str], duration_seconds: Optional[int]) -> dict:
    with SessionLocal() as s:
        m = s.get(Music, music_id)
        if not m:
            raise NotFound(f"music {music_id}")
        if title is not None:
            m.title = title
        if artist is not None:
            m.artist = artist
        if album is not None:
            m.album = album
        if duration_seconds is not None:
            m.duration_seconds = duration_seconds
        s.commit(); s.refresh(m)
        return _music_dto(m)


def delete_music(music_id: int) -> bool:
    with SessionLocal() as s:
        m = s.get(Music, music_id)
        if not m:
            return False
        s.delete(m); s.commit()
        return True


# =================== PLAYLISTS ===================
def create_playlist(name: str, user_id: int) -> dict:
    with SessionLocal() as s:
        if not s.get(User, user_id):
            raise NotFound(f"user {user_id}")
        p = Playlist(name=name, user_id=user_id)
        s.add(p); s.commit(); s.refresh(p)
        return _playlist_dto(p)


def get_playlist(playlist_id: int) -> dict:
    with SessionLocal() as s:
        p = s.get(Playlist, playlist_id)
        if not p:
            raise NotFound(f"playlist {playlist_id}")
        return _playlist_dto(p)


def list_playlists(limit: int = 100, offset: int = 0) -> list[dict]:
    with SessionLocal() as s:
        rows = s.scalars(
            select(Playlist).order_by(Playlist.id).limit(limit).offset(offset)
        ).all()
        return [_playlist_dto(p) for p in rows]


def update_playlist(playlist_id: int, name: Optional[str]) -> dict:
    with SessionLocal() as s:
        p = s.get(Playlist, playlist_id)
        if not p:
            raise NotFound(f"playlist {playlist_id}")
        if name is not None:
            p.name = name
        s.commit(); s.refresh(p)
        return _playlist_dto(p)


def delete_playlist(playlist_id: int) -> bool:
    with SessionLocal() as s:
        p = s.get(Playlist, playlist_id)
        if not p:
            return False
        s.delete(p); s.commit()
        return True


# =========== CONSULTAS DO ENUNCIADO (relações) ===========
def list_user_playlists(user_id: int) -> list[dict]:
    """Todas as playlists de um determinado usuário."""
    with SessionLocal() as s:
        rows = s.scalars(
            select(Playlist).where(Playlist.user_id == user_id)
            .order_by(Playlist.id)
        ).all()
        return [_playlist_dto(p) for p in rows]


def list_playlist_musics(playlist_id: int) -> list[dict]:
    """Todas as músicas de uma determinada playlist (ordenadas por posição)."""
    with SessionLocal() as s:
        rows = s.execute(
            select(Music).join(PlaylistMusic, PlaylistMusic.music_id == Music.id)
            .where(PlaylistMusic.playlist_id == playlist_id)
            .order_by(PlaylistMusic.position)
        ).scalars().all()
        return [_music_dto(m) for m in rows]


def list_playlists_with_music(music_id: int) -> list[dict]:
    """Todas as playlists que contêm uma determinada música."""
    with SessionLocal() as s:
        rows = s.execute(
            select(Playlist).join(PlaylistMusic, PlaylistMusic.playlist_id == Playlist.id)
            .where(PlaylistMusic.music_id == music_id)
            .order_by(Playlist.id)
        ).scalars().all()
        return [_playlist_dto(p) for p in rows]


def add_music_to_playlist(playlist_id: int, music_id: int,
                          position: int = 0) -> bool:
    with SessionLocal() as s:
        if not s.get(Playlist, playlist_id):
            raise NotFound(f"playlist {playlist_id}")
        if not s.get(Music, music_id):
            raise NotFound(f"music {music_id}")
        exists = s.get(PlaylistMusic, (playlist_id, music_id))
        if exists:
            return True
        s.add(PlaylistMusic(playlist_id=playlist_id, music_id=music_id,
                            position=position))
        s.commit()
        return True


def remove_music_from_playlist(playlist_id: int, music_id: int) -> bool:
    with SessionLocal() as s:
        res = s.execute(
            delete(PlaylistMusic).where(
                PlaylistMusic.playlist_id == playlist_id,
                PlaylistMusic.music_id == music_id,
            )
        )
        s.commit()
        return res.rowcount > 0


def counts() -> dict:
    with SessionLocal() as s:
        return {
            "users": s.scalar(select(func.count()).select_from(User)),
            "musics": s.scalar(select(func.count()).select_from(Music)),
            "playlists": s.scalar(select(func.count()).select_from(Playlist)),
        }

File: api-comparison/typescript/src/common/repository.ts
/**
 * Repositório com TODA a lógica de negócio — espelho fiel de
 * common/repository.py (Python). Cada rota do server.ts é apenas um adaptador
 * fino sobre estas funções, exatamente como no lado Python. Mesma semântica de
 * NotFound, mesmas consultas do enunciado.
 */
import { pool } from "./db.js";

export class NotFound extends Error {}

export interface UserDTO { id: number; name: string; email: string; }
export interface MusicDTO {
  id: number; title: string; artist: string;
  album: string | null; duration_seconds: number;
}
export interface PlaylistDTO { id: number; name: string; user_id: number; }

// =================== USERS ===================
export async function createUser(name: string, email: string): Promise<UserDTO> {
  const r = await pool.query(
    "INSERT INTO users (name, email) VALUES ($1, $2) RETURNING id, name, email",
    [name, email],
  );
  return r.rows[0];
}

export async function getUser(id: number): Promise<UserDTO> {
  const r = await pool.query("SELECT id, name, email FROM users WHERE id = $1", [id]);
  if (r.rowCount === 0) throw new NotFound(`user ${id}`);
  return r.rows[0];
}

export async function listUsers(limit = 100, offset = 0): Promise<UserDTO[]> {
  const r = await pool.query(
    "SELECT id, name, email FROM users ORDER BY id LIMIT $1 OFFSET $2",
    [limit, offset],
  );
  return r.rows;
}

export async function updateUser(
  id: number, name?: string | null, email?: string | null,
): Promise<UserDTO> {
  const r = await pool.query(
    `UPDATE users SET
       name  = COALESCE($2, name),
       email = COALESCE($3, email)
     WHERE id = $1
     RETURNING id, name, email`,
    [id, name ?? null, email ?? null],
  );
  if (r.rowCount === 0) throw new NotFound(`user ${id}`);
  return r.rows[0];
}

export async function deleteUser(id: number): Promise<boolean> {
  const r = await pool.query("DELETE FROM users WHERE id = $1", [id]);
  return (r.rowCount ?? 0) > 0;
}

// =================== MUSICS ===================
export async function createMusic(
  title: string, artist: string, album: string | null, durationSeconds = 0,
): Promise<MusicDTO> {
  const r = await pool.query(
    `INSERT INTO musics (title, artist, album, duration_seconds)
     VALUES ($1, $2, $3, $4)
     RETURNING id, title, artist, album, duration_seconds`,
    [title, artist, album, durationSeconds],
  );
  return r.rows[0];
}

export async function getMusic(id: number): Promise<MusicDTO> {
  const r = await pool.query(
    "SELECT id, title, artist, album, duration_seconds FROM musics WHERE id = $1", [id],
  );
  if (r.rowCount === 0) throw new NotFound(`music ${id}`);
  return r.rows[0];
}

export async function listMusics(limit = 100, offset = 0): Promise<MusicDTO[]> {
  const r = await pool.query(
    `SELECT id, title, artist, album, duration_seconds
     FROM musics ORDER BY id LIMIT $1 OFFSET $2`,
    [limit, offset],
  );
  return r.rows;
}

export async function updateMusic(
  id: number, title?: string | null, artist?: string | null,
  album?: string | null, durationSeconds?: number | null,
): Promise<MusicDTO> {
  const r = await pool.query(
    `UPDATE musics SET
       title            = COALESCE($2, title),
       artist           = COALESCE($3, artist),
       album            = COALESCE($4, album),
       duration_seconds = COALESCE($5, duration_seconds)
     WHERE id = $1
     RETURNING id, title, artist, album, duration_seconds`,
    [id, title ?? null, artist ?? null, album ?? null, durationSeconds ?? null],
  );
  if (r.rowCount === 0) throw new NotFound(`music ${id}`);
  return r.rows[0];
}

export async function deleteMusic(id: number): Promise<boolean> {
  const r = await pool.query("DELETE FROM musics WHERE id = $1", [id]);
  return (r.rowCount ?? 0) > 0;
}

// =================== PLAYLISTS ===================
export async function createPlaylist(name: string, userId: number): Promise<PlaylistDTO> {
  const u = await pool.query("SELECT 1 FROM users WHERE id = $1", [userId]);
  if (u.rowCount === 0) throw new NotFound(`user ${userId}`);
  const r = await pool.query(
    "INSERT INTO playlists (name, user_id) VALUES ($1, $2) RETURNING id, name, user_id",
    [name, userId],
  );
  return r.rows[0];
}

export async function getPlaylist(id: number): Promise<PlaylistDTO> {
  const r = await pool.query("SELECT id, name, user_id FROM playlists WHERE id = $1", [id]);
  if (r.rowCount === 0) throw new NotFound(`playlist ${id}`);
  return r.rows[0];
}

export async function listPlaylists(limit = 100, offset = 0): Promise<PlaylistDTO[]> {
  const r = await pool.query(
    "SELECT id, name, user_id FROM playlists ORDER BY id LIMIT $1 OFFSET $2",
    [limit, offset],
  );
  return r.rows;
}

export async function updatePlaylist(id: number, name?: string | null): Promise<PlaylistDTO> {
  const r = await pool.query(
    `UPDATE playlists SET name = COALESCE($2, name)
     WHERE id = $1 RETURNING id, name, user_id`,
    [id, name ?? null],
  );
  if (r.rowCount === 0) throw new NotFound(`playlist ${id}`);
  return r.rows[0];
}

export async function deletePlaylist(id: number): Promise<boolean> {
  const r = await pool.query("DELETE FROM playlists WHERE id = $1", [id]);
  return (r.rowCount ?? 0) > 0;
}

// =========== CONSULTAS DO ENUNCIADO (relações) ===========
/** Todas as playlists de um determinado usuário. */
export async function listUserPlaylists(userId: number): Promise<PlaylistDTO[]> {
  const r = await pool.query(
    "SELECT id, name, user_id FROM playlists WHERE user_id = $1 ORDER BY id",
    [userId],
  );
  return r.rows;
}

/** Todas as músicas de uma determinada playlist (ordenadas por posição). */
export async function listPlaylistMusics(playlistId: number): Promise<MusicDTO[]> {
  const r = await pool.query(
    `SELECT m.id, m.title, m.artist, m.album, m.duration_seconds
     FROM musics m
     JOIN playlist_musics pm ON pm.music_id = m.id
     WHERE pm.playlist_id = $1
     ORDER BY pm.position`,
    [playlistId],
  );
  return r.rows;
}

/** Todas as playlists que contêm uma determinada música. */
export async function listPlaylistsWithMusic(musicId: number): Promise<PlaylistDTO[]> {
  const r = await pool.query(
    `SELECT p.id, p.name, p.user_id
     FROM playlists p
     JOIN playlist_musics pm ON pm.playlist_id = p.id
     WHERE pm.music_id = $1
     ORDER BY p.id`,
    [musicId],
  );
  return r.rows;
}

export async function addMusicToPlaylist(
  playlistId: number, musicId: number, position = 0,
): Promise<boolean> {
  const p = await pool.query("SELECT 1 FROM playlists WHERE id = $1", [playlistId]);
  if (p.rowCount === 0) throw new NotFound(`playlist ${playlistId}`);
  const m = await pool.query("SELECT 1 FROM musics WHERE id = $1", [musicId]);
  if (m.rowCount === 0) throw new NotFound(`music ${musicId}`);
  await pool.query(
    `INSERT INTO playlist_musics (playlist_id, music_id, position)
     VALUES ($1, $2, $3)
     ON CONFLICT (playlist_id, music_id) DO NOTHING`,
    [playlistId, musicId, position],
  );
  return true;
}

export async function removeMusicFromPlaylist(
  playlistId: number, musicId: number,
): Promise<boolean> {
  const r = await pool.query(
    "DELETE FROM playlist_musics WHERE playlist_id = $1 AND music_id = $2",
    [playlistId, musicId],
  );
  return (r.rowCount ?? 0) > 0;
}

export async function counts(): Promise<{ users: number; musics: number; playlists: number; }> {
  const r = await pool.query(
    `SELECT
       (SELECT COUNT(*) FROM users)     AS users,
       (SELECT COUNT(*) FROM musics)    AS musics,
       (SELECT COUNT(*) FROM playlists) AS playlists`,
  );
  const row = r.rows[0];
  return {
    users: Number(row.users),
    musics: Number(row.musics),
    playlists: Number(row.playlists),
  };
}
